[yeon] 미션3: 체스판 초기화

kimnayeon0108 opened the pull request:
### 미션3: 체스판 초기화

안녕하세요! 미션3 체스판 초기화 구현하고 리뷰 요청 드립니다.
아래에 각각 클래스의 필드와 메소드에 대한 설명을 기술하였으니 참고 부탁드립니다!🙇‍♀️

1. **Borad 클래스**
- 리스트  pawns를 삭제하고 pawn의 색깔에 따라 해당되는 리스트에 담을 수 있도록 새로운 리스트 whitePawns 와 blackPawns를 생성했습니다.
- distinguish() 메소드 추가 : 해당 pawn이 white인지 black인지 구별해서 해당하는 컬러의 리스트를 반환하는 메소드를 추가했습니다.
    → add(), size(), findPawn() 메소드 호출 시 distiguish() 메소드를 호출해서 whitePawns와 blackPawns 두 리스트 중 어느 리스트에 담겨야 할지 구별하는 기능을 추가했습니다.  
- 상수 PAWNSNUM와 COLUMN 추가 : 각 색깔의 pawn 개수(8)와 체스판의 열의 개수(8)를 값으로 갖는 PAWNSNUM과 COLUMN을 추가했습니다.
- initialize() 메소드 : PAWNSNUM 만큼 white, black 폰을 생성하여, 해당되는 리스트에 넣어주는 기능을 합니다.
- getPawnResult() 메소드 : 인자로 List를 받아서 리스트의 요소들의 representation(폰을 표현하는 알파벳)을 하나의 String으로 만들어주는 역할을 합니다.
- getWhitePawnResult() 메소드와 getBlackPawnResult() : 내부적으로 getPawnResult()를 호출하여 갖고 있는 요소들을 하나의 String으로 반환합니다.
- getEmptyResult() 메소드 : 체스판의 피스가 놓여있지 않은 열은 "."으로 이루어진 String을 반환하는 기능을 추가했습니다.
- print() 메소드 : 콘솔화면에 체스판을 출력합니다. getBlackPawnResult(), getWhitePawnResult(), getEmptyResult()를 호출해서 각 열마다 해당되는 폰이나 "."을 출력하도록 했습니다.

2. **Chess 클래스**
- main메소드를 갖는 클래스 입니다.
- Scanner를 이용해서 시작과 종료 기능을 구현했습니다.
- 사용자에게 입력을 받아서 시작과 종료를 하는 기능, Board 객체를 생성해서 board를 초기화, 출력하는 기능들도 따로 메소드를 생성해서 분리해야 하나 싶었는데 이런식으로 하는 것이 나을지 아니면 현재 코드처럼 메인메소드에 다 넣는 것이 나을지 잘 모르겠어서 일단 현재 상태로 리뷰 요청드립니다!

3. **Pawn 클래스**
- WHITE와 BLACK을 WHITE_COLOR와 BLACK_COLOR라고 상수명을 변경하여 색깔이라는 의미를 더 명확하게 하였습니다.
- 상수 WHITE_REPRESENTATION과 BLACK_REPRESENTATION 을 추가하여 각 폰에 해당되는 알파벳 p와 P를 대입하였습니다.
- 전역변수 representation을 추가하여 폰을 묘사하는 알파벳을 갖도록 했습니다.
- 생성자 수정 : 인자로 color뿐만아니라 representation를 추가하였습니다.

4. **BoardTest 클래스**
- setUp() 메소드 : Pawn 인스턴스 생성시 color, representation 두개의 인자를 받도록 생성자를 변경해서 이에 맞게 코드를 변경했습니다.
- create() 메소드 : Board클래스에서 whitePawns와 blackPawns 리스트가 따로 구현되어 있어서 이에 맞게 코드를 수정했습니다.
- print() 메소드 : Board클래스의 getEmptyResult()가 잘 동작하는지 확인하는 테스트 코드를 추가했습니다.
- check() 메소드 : Board클래스의 size(), findPawn()메소드를 변경해서 이에 맞게 코드를 수정했습니다.

5. **PawnTest 클래스**
- Pawn의 객체를 생성할때 인자로 넣은 값으로 color와 representation이 잘 초기화되었는지 확인하도록 하였습니다.

바쁜시간에 시간내서 리뷰해주셔서 감사합니다😊

kimnayeon0108 committed:
Add dependency in  build.gradle

kimnayeon0108 committed:
Change Test code in PawnTest class according to requirements

kimnayeon0108 committed:
Add representation variable and getRepresentation() method in Pawn class, add initialize() method in Board class

kimnayeon0108 committed:
Modify Board.java to merge

kimnayeon0108 committed:
Add whitePawns, blackPawns List and distinguish() method and modify add(), size(), findPawn() method

kimnayeon0108 committed:
Modify create() test code in BoardTest class

kimnayeon0108 committed:
Complete initialize() method in Board class

kimnayeon0108 committed:
Add getPawnResult() method to get rid of duplicated code

kimnayeon0108 committed:
Delete useless code

kimnayeon0108 committed:
Add print() method and COLUMN constant

kimnayeon0108 committed:
Add getEmptyResult() method

kimnayeon0108 committed:
Add print() test method to make sure getEmptyResult() method in Board class works

kimnayeon0108 committed:
Delete creating Board instance code, considered to be useless to check getEmptyResult() method works

kimnayeon0108 committed:
Delete useless code

kimnayeon0108 committed:
Add Chess.java file to create main method and start, end function by getting input from user

kimnayeon0108 committed:
Refactor code

kimnayeon0108 committed:
Refactor

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,20 +1,86 @@
-        this.pawns.add(pawn);
+    private static final int PAWNSNUM = 8;
+    private static final int COLUMN = 8;
+    private List<Pawn> whitePawns = new ArrayList<>(8);
+    private List<Pawn> blackPawns = new ArrayList<>(8);
```
`int` 파라메터를 갖는 생성자를 사용하셨네요. 좋습니다. 👍 
그렇다면 **생성자 파라메터 없이 어레이리스트를 생성한다면** 기본으로 갖는 용량 값이 얼마일까요?
한번 찾아보세요.

kimnayeon0108 replied:
@wheejuni 생성자 파라메터 없이 생성하면 capacity가 10인 어레이리스트가 생성됩니다. 검색해보니 ArrayList에는 size와 capacity로 크기에 관한 개념(?)이 따로 있어서 이부분을 더 학습해봐야할 것 같습니다~

wheejuni replied:
@kimnayeon0108 좋네요. 그렇다면 초기에 생성한 용량이 가득 찼을때의 어레이리스트 동작은 어떨까요?

예를 들어, 캐퍼시티를 8로 부여한 어레이리스트에서 여덟개의 원소가 가득 채워졌다면, 어떻게 될까요? 이때의 동작도 한번 학습해보세요.

kimnayeon0108 replied:
@wheejuni capacity가 가득 찼을때 어레이리스트는 기존 캐퍼시티의 1.5배 + 1의 크기로 증가되고 
초기용량을 8로 설정한 어레이리스트에서 여덟개의 원소가 가득 채워지고, 아홉번째 원소가 추가될 경우 capacity가 13인 배열을 내부적으로 새로 생성합니다.

kyupid replied:
`ArrayList(8)` 이라고하면 단지 size가 8이되는 건줄 알았는데.. size외에 capacity 라는 개념이 있었군요. 저도 학습해봐야겠습니다.

kyupid replied:
공부하고 정리한 내용을 공유합니다. 틀린 내용이 있으면 꼭 고쳐주시면 감사하겠습니다.

---

ArrayList를 생성자 없이 그냥 생성하면 capacity 가 10이다. 그런데 i를 직접 지정해주면 initial capacity 를 지정할 수 있다.

그런데 여기서 size 와 capacity 는 무엇이 다른걸까? size는 흔히 사용해왔듯이 add한 데이터의 길이를 나타낸다. capacity는 그 단어의 어감이 최대로 넣을 수 있는 데이터의 upper limit 이라고 생각할 수 있는데, 단지 얼마나 많은 데이터를 현재 ArrayList에 담을 수있는 지를 나타내는 값일 뿐이다.

ArrayList를 배울 때 배열처럼 길이를 지정해주는 게 아니라 데이터를 넣을 때 유동적으로 바뀐다는 것을 배웠다. 지금 생각해보면 분명 ArrayList 를 선언할 때 기본적으로 용량이 정해져있을 것이다. 근데 그게 999999999는 아닐 것이다. 메모리 남용이니까. 그런 개념으로 capacity가 있는 것이다. 처음에 일정량 들을어 올수 있는 값을 정해놓고 꽉 채워지면 내부적으로 정해진 룰에 따라 (1.5배 + 1) 용량을 계속 늘려나가는 것이다. 

API 에서 ArrayList를 제공하는 인터페이스(List)는 기본적으로 capacity 를 사용하는 것과 같은 기능을 제공하지 않는다. 주요한 이유 중 하나는 사용자가 이것을 쓸때없이 신경쓰지 않도록 하기 위함이다. 왜냐면 capacity 같은 추상적인 개념에서 벗어나 ArrayList는 무제한 배열(unbounded array)로 생각되어져야 하기 때문이다.

> [Capacity of ArrayList](https://stackoverflow.com/questions/3564837/capacity-of-arraylist)

wheejuni replied:
@kyu-kim-kr `ArrayList` 가 `Array List` 인 이유가 있겠죠. 
내부적으로 배열을 사용합니다. 말씀하신 것처럼 처음부터 너무 큰 배열을 선언하면 메모리 낭비가 되니까, 필요한 최소한을 만들어두고 계속된 `.add()` 로 용량이 증대될 필요가 있을 경우 그때 비로소 더 넓은 배열을 생성하고, 이사가는 것입니다. 
이때 _더 넓은 배열을 생성하고 이사가는 작업_ 은 컴퓨터의 연산을 다소 필요로 합니다. 자동으로 관리되도록 내버려둬도 괜찮을텐데, 굳이 `ArrayList` 클래스의 생성자로 capacity를 지정할 수 있게 한 것은, 개발자의 책임과 판단에 따라 큰 용량이 처음부터 필요할 리스트 객체에 대해선 미리 넓은 공간을 확보할 수 있게 하여, 연산력 소모를 아낄 수 있는 개발 방법을 제공하기 위해 그렇게 한 것입니다.

tmdgusya commented on src/main/java/chess/Board.java:
```diff
@@ -3,46 +3,60 @@
+        if (pawn.getColor().equals(Pawn.BLACK_COLOR)) {
+            return blackPawns;
+        }
+        return null;
```
지금 코드를 보면 whitePawns 아니면 blackPawns List 를 리턴해주는 코드인데, 이럴 경우는  A 혹은 B 로 케이스가 정확하므로 아래와 같은 방식이 좋은 방식이 아닐까 합니다 위의 if 문을 거치지않게되면 무조건 blackPawns 를 리턴하는 식으로요. 만약 white 라면 앞의 whitePawns list 가 리턴되고 끝날테니까요!

```java
if (pawn.getColor().equals(Pawn.WHITE_COLOR)) {
            return whitePawns;
        }
return blackPawns;
```

kimnayeon0108 replied:
@tmdgusya 오 훨씬 좋은방법 같아요! 감사합니다 ㅎㅎ 참고해서 코드 수정해볼게요~

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,20 +1,86 @@
+    private List<Pawn> whitePawns = new ArrayList<>(8);
+    private List<Pawn> blackPawns = new ArrayList<>(8);
+
+    public List distinguish(Pawn pawn) {
```
타입 파라메터는 필요 없나요?
```suggestion
    public List<Pawn> distinguish(Pawn pawn) {

kimnayeon0108 replied:
@wheejuni 이부분을 List<Pawn>으로 할 생각을 못했습니다ㅠㅠ distinguish()를 호출하는 findPawn 메소드에서 return 부분에서 (Pawn)으로 형변환을 안해주면 컴파일 에러가 뜨고, 이유를 몰랐는데 distinguish 메소드의 리턴타입에 타입 파라메터를 추가하지 않아서 발생한 문제라는 것을 알았습니다. 수정하도록 하겠습니다!

tmdgusya commented on src/main/java/chess/Board.java:
```diff
@@ -3,46 +3,60 @@
+    }
+
+    public Pawn findPawn(Pawn pawn, int idx) {
+        return (Pawn) distinguish(pawn).get(idx);
```
이렇게 Type 변환을 하기보다는 distinguish 메소드 자체에 return 타입에 List < Pawn > 으로 지정해주는 것이 좋아보입니다

kimnayeon0108 replied:
@tmdgusya 이부분에서 형변환을 안해주면 컴파일에러가 뜨고 이유를 몰라서 의아했는데 distinguish메소드 리턴타입에 List<Pawn>으로 해주면 되는 거였군요.... 감사합니다!👍🙇‍♀️

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,20 +1,86 @@
+    }
+
+    public void add(Pawn pawn) {
+        List list = distinguish(pawn);
```
`distinguish()` 는 여기 밖에서 사용될 필요가 거의 없어보이는데요,
그렇다면 `distinguish()` 를 `private` 으로 바꾸는 건 어떨까요?

kimnayeon0108 replied:
@wheejuni 네엡 수정하도록 하겠습니다!

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,20 +1,86 @@
+        }
+    }
+
+    public void print() {
```
이 메소드가 `println` 까지 하는 책임을 갖지는 않는 게 좋겠네요.
`System.out.println()` 은 메인 메소드에서 호출되게 다시 구현해주세요.

kimnayeon0108 replied:
@wheejuni 네에 String을 리턴하도록 수정해보겠습니다!

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,20 +1,86 @@
+        for (int i = 0; i < COLUMN; i++) {
+            if (i == 1) {
+                result.append(getBlackPawnResult());
+            } else if (i == 6) {
```
`else if` 사용하지 않도록 개선해주세요.

kimnayeon0108 replied:
@wheejuni 네엡 코드 수정하겠습니다~

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,20 +1,86 @@
 
-    public Pawn findPawn(int idx) {
-        return pawns.get(idx);
+    public String getPawnResult(List<Pawn> pawns) {
```
이 메소드도 `private` 이어도 충분할 것 같네요.

wheejuni commented on src/main/java/chess/pieces/Pawn.java:
```diff
@@ -1,21 +1,30 @@
-    public Pawn(){
-        this(WHITE);
+    public Pawn(String color, char representation) {
+        this.color = color;
+        this.representation = representation;
```
`color` 가 `representation` 까지 가질 수 있는 방법은 없을까요?

kimnayeon0108 replied:
@wheejuni 음.. 그럼 enum을 선언해서 각 컬러마다 representation을 갖게 하도록 구현해보면 될거 같습니다. enum으로 시도해보겠습니다!

kimnayeon0108 replied:
git 명령어 수행하다가 오랫동안 헤매서 아직 이부분은 수정하지 못했습니다. 내일 바로 수정하도록 하겠습니다!

wheejuni reviewed (changes requested):
3단계 진행하시느라 정말 고생 많으셨어요. 💯 
많은 고민을 하고 코드를 작성하신 티가 납니다. 다만 개선될 여지도 있어보입니다. 
개선 방향을 코멘트를 통해 제안해 드립니다. 검토해보시고, 반영할 수 있는 부분은 반영해주시고, 추가적인 질문이 생기신다면 언제든지 대댓글 남겨주세요! 👍

kimnayeon0108 committed:
Change return type of distinguish method()

kimnayeon0108 committed:
Update src/main/java/chess/Board.java

Co-authored-by: Hwijun Jeong <me@wheejuni.com>

kimnayeon0108 committed:
Modify Distinguish() method

kimnayeon0108 committed:
Modify print(), getPawnResult() method

kimnayeon0108 committed:
Merge branch 'kimnayeon0108' of https://github.com/kimnayeon0108/java-chess into step3-initialize-board

kimnayeon0108 committed:
Merge origin branch 'step3-initialize-board'

kimnayeon0108 committed:
Add enum Color
contains representation field and getter method

kimnayeon0108 commented:
@wheejuni 리뷰해주신 부분 전부 수정 완료했습니다~!

wheejuni commented on src/main/java/chess/pieces/Color.java:
```diff
@@ -0,0 +1,15 @@
+package chess.pieces;
+
+public enum Color {
+    WHITE('p'), BLACK('P');
```
`enum`을 선언할때는 한 줄에 하나씩 선언하는 것을 원칙으로 해 주세요.

```suggestion
    WHITE('p'),
    BLACK('P');

kimnayeon0108 replied:
네에 수정 완료했습니다!

wheejuni commented on src/main/java/chess/pieces/Pawn.java:
```diff
@@ -1,21 +1,25 @@
-    public static final String BLACK = "black";
-    private String color;
+    private Color color;
+    private char representation;
```
음? 좀전에 `Color` 에서 `representation` 을 관리하기로 구현하지 않았나요?

kimnayeon0108 replied:
@wheejuni 아 그렇네요 Pawn에서도 representation을 갖고 있어야한다고 생각했는데 다시생각해보니 Color에서 갖고 있으니 이부분은 없애도 되겠군요 감사합니다!

wheejuni reviewed (changes requested):
코드 수정하시느라, 그리고 제 추가질문에 답해주시느라 고생 많으셨습니다 💯 
다만 간단히 짚어볼 부분이 있어 다시 리뷰 의견 드립니다.
확인해주세요~

kimnayeon0108 committed:
Remove representation variable
Color has representation, so remove duplicated property

kimnayeon0108 committed:
Remove getRepresentation() method in Pawn.java
Color has the same method and fix other methods that include getRepresentation() method

kimnayeon0108 committed:
Fix enum Color
enum constants should be declared one per line

kyupid commented on src/main/java/chess/Board.java:
```diff
@@ -1,20 +1,62 @@
+    private static final int COLUMN = 8;
+    private List<Pawn> whitePawns = new ArrayList<>(8);
+    private List<Pawn> blackPawns = new ArrayList<>(8);
+    private List<List<Pawn>> board = new ArrayList(8);
```
@wheejuni 저도 이렇게 구현하고 싶었던 것 같습니다.

kyupid commented on src/main/java/chess/Board.java:
```diff
@@ -1,20 +1,86 @@
+        return blackPawns;
+    }
+
+    public void add(Pawn pawn) {
+        List list = distinguish(pawn);
+        list.add(pawn);
+    }
```
궁금한게 있는데요~~ 새로운 list에다가  distinguish(pawn) 하면 whitePawns 나 blackPawns 가 들어가잖아요?

이 때 만약에 whitePawns 가 들어갔으면 `list` 는 새로 생성되는 게 아니라 기존에 있던 whitePawns의 주소를 가리키게 되는건가요?

그런 다음에 그 list.add(pawn) 하면 기존에 있던 whitePawns 에도 pawn 이 들어가는거죠?

kyupid replied:
그리고 작성하신 size() 나 findPawn() 처럼  `distinguish(pawn).add(pawn);` 이렇게 작성해도 되려나요?

kimnayeon0108 replied:
@kyu-kim-kr kyu! 새로운 list에다가 distinguish(pawn)하는게 무슨 뜻인지 잘 모르겠어요ㅠㅠ 혹시 다시 설명해주실수 있나요...?  이 메소드는 pawn 객체를 넣으면 어떤 리스트에 들어가야 하는지를 반환해주는 용도로 작성했습니당

kimnayeon0108 replied:
@kyu-kim-kr add()메소드도 size()나 findPawn()처럼 distinguish(pawn).add(pawn); 이렇게 작성해도 됩니다!

kimnayeon0108 replied:
@kyu-kim-kr 아 제가 질문을 잘못 이해했네요 말씀하신대로 기존에 있던 whitePawns에 add됩니다!

wheejuni reviewed (approved):
👍
